### ISSUE TITLE
Proposal: Add option to pass in state for each API request

### ISSUE THREAD

**Issue by Cidan** [nicklaw5/helix#128]
First of all, thank you for your amazing work on this library, it's great!

The library in general is written with single-user use cases in mind and doesn't work well in a multi-tenant system, i.e. a microservice which handles API calls for multiple users, bot id's, etc. This is because the state of all authentication mechanisms is stored within the instantiated object it self.

One option would be to create a new object per user, but this creates overhead, eliminates the possibility of some design patterns (i.e. singleton instantiation), and would require a tighter coupling of the driver to the authentication source.

Ideally, all authentication would be stateless and be passed in per call. At the very minimum, user and application tokens being passed in per API call would provide sufficient abstraction insofar that multiple users can be used on the same bot, concurrently, without having to create a new object for every user.

I'm curious to hear what you think @nicklaw5 about this proposal.

Thanks!

**Comment by nicklaw5** [comment]
I agree that is generally a good idea on the surface, but without diving in and having a deeper look I'm not sure what changes need to be implemented.

I'm also rather busy at the moment and not sure when I'll be able to have a look at this.

**Comment by Cidan** [comment]
I'll take a look and outline what changes would need to be done in order to get this in place. At the very least, we should be able to measure the level of effort here.

**Comment by graytonio** [comment]
I'm running into similar issues with wanting to have multitenant applications has this feature been abandoned?
